indexer-balances/src/db_adapters/balance_changes.rs

```diff
@@ -624,51 +708,21 @@ async fn get_balance_retriable(
 async fn get_balance(
     account_id: &near_indexer_primitives::types::AccountId,
     block_hash: &near_indexer_primitives::CryptoHash,
-    balance_cache: &crate::BalanceCache,
     json_rpc_client: &near_jsonrpc_client::JsonRpcClient,
 ) -> anyhow::Result<crate::BalanceDetails> {
-    let mut balances_cache_lock = balance_cache.lock().await;
-    let result = match balances_cache_lock.cache_get(account_id) {
-        None => {
-            let account_balance =
-                match get_account_view(json_rpc_client, account_id, block_hash).await {
-                    Ok(account_view) => Ok(crate::BalanceDetails {
-                        non_staked: account_view.amount,
-                        staked: account_view.locked,
-                    }),
-                    Err(err) => match err.handler_error() {
-                        Some(RpcQueryError::UnknownAccount { .. }) => Ok(crate::BalanceDetails {
-                            non_staked: 0,
-                            staked: 0,
-                        }),
-                        _ => Err(err.into()),
-                    },
-                };
-            if let Ok(balance) = account_balance {
-                balances_cache_lock.cache_set(account_id.clone(), balance);
-            }
-            account_balance
-        }
-        Some(balance) => Ok(*balance),
-    };
-    drop(balances_cache_lock);
-    result
-}
-
-async fn save_latest_balance(
-    account_id: near_indexer_primitives::types::AccountId,
-    balance: &crate::BalanceDetails,
-    balance_cache: &crate::BalanceCache,
-) {
-    let mut balances_cache_lock = balance_cache.lock().await;
-    balances_cache_lock.cache_set(
-        account_id,
-        crate::BalanceDetails {
-            non_staked: balance.non_staked,
-            staked: balance.staked,
+    match get_account_view(json_rpc_client, account_id, block_hash).await {
```

Comment:
This will not work, unfortunately.
RPC can return only the balance for the end of the block. The user balance can change many times during one block, and we need to be careful applying all the deltas in the right order.

Reply:
Thanks for the insight 👍🏼 

If we need to process sequentially, to ensure that account changes within the block are captured correctly, can we group by `account_id` and process each of those in parallel?

We could update the types to:

```rust
pub struct BalanceDetails {
    pub non_staked: near_indexer_primitives::types::Balance,
    pub staked: near_indexer_primitives::types::Balance,
}

struct AccountChangesBalances {
    pub validators: Vec<crate::BalanceDetails>,
    pub transactions: HashMap<near_indexer_primitives::CryptoHash, crate::BalanceDetails>,
    pub receipts: HashMap<near_indexer_primitives::CryptoHash, crate::BalanceDetails>,
    pub rewards: HashMap<near_indexer_primitives::CryptoHash, crate::BalanceDetails>,
}

type ChangesByAccount = HashMap<near_indexer_primitives::types::AccountId, AccountChangesBalances>;
```

Then for each account with `ChangesByAccount`, we could spawn a new task which would essentially call a modified `store_changes_for_chunk()`, which works through validator/transaction/receipt changes for each account in sequence.

We'd need to update the cache slightly so that each account could read/write separately, without blocking other processes. But I'm not sure if this will impact the `Inbound` events, i.e. when we create a `NearBalanceEvent` for `involved_account_id`. We couldn't guarantee that that accounts balance would be cached and it would therefore always fallback to RPC. Do you know if this would be an issue?

Reply:
By "process", you mean insert to the DB? If so - yes, your approach should work.
Grouping by account id may be trickier than expected, because you may find balance changing event for account_id_1 while processing the data from account_id_2. So this preprocessing can't be easily parallelised.

In the ideal world, after some time, the number of RPC calls should decrease dramatically, we should have almost everything in cache. 
But I don't remember whether we make RPC calls to check if our calculations are correct.